### PR TITLE
Allow including master table in cascading delete

### PR DESCRIPTION
**Pull request by CBroz1**
#151 included a discussion of 'Option 2' in which the master would be deleted during a cascade, with a possible solution proposed in this PR. This feature would be particularly useful to the Spyglass team because our ['Merge tables'](https://lorenfranklab.github.io/spyglass/0.5/misc/merge_tables/) make use of foreign-key references in parts to version pipelines. 

Happy to incorporate feedback

### CODE REVIEW

**Comment by lfrank** [comment]
@dimitri-yatsenko @ttngu207 Hi Dmitri and Thinh. I just wanted to check in on this pull request and another one (https://github.com/datajoint/datajoint-python/pull/1160). It would really help us to get these into a new release if that's possible.

**Comment by ethho** [comment]
> it would be helpful to know for our (Spyglass) planning purposes whether this PR is likely to get reviewed/merged in the near future.

Hi @edeno, I'll review this by end of week. Thanks all for the PR.

@ttngu207 @dimitri-yatsenko @lfrank Do we need to publish a new release? Or is it sufficient to merge to `main`?

**Comment by edeno** [comment]
It would be greatly helpful for us to have a release.

File: datajoint/table.py [hunk 1]
```diff
@@ -486,6 +486,7 @@ def delete(
         transaction: bool = True,
         safemode: Union[bool, None] = None,
         force_parts: bool = False,
+        include_parts: bool = True,
```
Comment (ttngu207) [inline datajoint/table.py:489]: I'd suggest a different argument name. It's not super obvious what `include_parts` means, as ones must expect that cascade delete will also delete the parts.

Here, we mean to include also the "master" table(s) of the part table(s) being deleted.

Suggesting to rename to something like `force_masters` (for consistency with the arg `force_parts`)
Or perhaps consider using an environment variable or configuration (in `dj.config`) to control for this behavior (similar to `DJ_SUPPORT_FILEPATH_MANAGEMENT`)

Comment (dimitri-yatsenko) [reply datajoint/table.py:489]: I would prefer to have a single consistent behavior, not making it configurable.

**Comment by ethho** [comment]
@CBroz1 I would recommend checking out changes from #1164 so that tests run.

**Comment by CBroz1** [comment]
> @CBroz1 I would recommend checking out changes from #1164 so that tests run.

Thanks @ethho! If you plan to merge that first, I'll keep an eye out and fetch upstream when it does

**Comment by ethho** [comment]
> If you plan to merge that first, I'll keep an eye out and fetch upstream when it does

@CBroz1 #1164 was merged to `master`.

**Comment by CBroz1** [comment]
Thanks, @ethho. I fetched from upstream

**Review by ethho — CHANGES_REQUESTED** [review]
#### Summary

Looks good overall. I request the following changes:

1. Update documentation to reflect the new behavior. It is sufficient to update [this section on deletes](https://github.com/datajoint/datajoint-python/blob/685b42ccbc85c4b03a347311e272539cfc368f5e/docs/src/manipulation/delete.md#L23).
2. Rename kwarg to `force_masters` as discussed above
3. Expand regression tests to cover edge cases

#### Edge Cases

The logic changes in `delete` are sensible and address the requirements laid out in #151.

While I can't think of any breaking edge cases, @ttngu207 and I think that we should include some of these cases in regression tests.
Notably, we should test what happens when a part table is visited more than once in the cascade.
The new tests cover the case where the master table is visited twice, and we save some work by skipping `visited_masters`.
But what if a part table shares a master with its children or grandchildren?
`cascade` will visit the downstream part table twice.
This is harmless currently because `quick_delete` is idempotent, but it's worth testing in case delete behavior changes. For example, we could add a table to `schema_simple`:

```python
class I(dj.Part):
    definition = """ # test no additional fk reference
    -> E
    id_i :int
    ---
    -> H
    """
```

#### Kwarg Interaction

No change requested, but note that `my_part_table.delete(force_parts=False, force_masters=True)` will delete the master table, even though `force_parts=False`. This behavior makes sense to me: `force_masters=True` will always supercede `force_parts`.

**Comment by CBroz1** [comment]
Thanks for your review @ethho!

> But what if a part table shares a master with its children or grandchildren?

Sure, happy to test this case. It looks like there are issues with a part being a child of a another part of the same master, given the tests failing in my latest commit. I'll spend some time tomorrow looking at some alternatives, but let me know if you have other suggestions

> `my_part_table.delete(force_parts=False, force_masters=True)` will delete the master table, even though `force_parts=False`. 

Given that this is currently the default behavior, should `force_parts=True` turn off `force_masters`, or log a warning? Otherwise, in order to get the previous `force_parts=True` behavior, users will need to set `force_parts=True, force_masters=False`.

**Comment by ethho** [comment]
> Given that this is currently the default behavior, should force_parts=True turn off force_masters, or log a warning?

@CBroz1 Based on my reading, `force_masters=True` always implies `force_parts=True`, since [master is always in deleted in this expression](https://github.com/datajoint/datajoint-python/blob/685b42ccbc85c4b03a347311e272539cfc368f5e/datajoint/table.py#L609) when `force_masters=True`. `force_masters=False` restores the behavior before this PR, where the value of `force_parts` is relevant.

Then we should document the behavior well and keep the logic as is, without raising a warning. A warning would get very annoying as the default behavior. One option that I don't endorse is: change default behavior to `force_masters=True, force_parts=True`, but this would result in surprising behavior for users accustomed to `delete()` who refactor to `delete(force_masters=False)` in the new version.

**Comment by CBroz1** [comment]
Thanks @ethho, sounds good. I made note of this in the changelog. 

If we could get this into a `14.2` release, my team would really benefit both from this PR as well as #1106. Happy to make adjustments here if that makes sense, but I understand if there are other features you're targeting for inclusion in `14.2`

**Comment by ethho** [comment]
> If we could get this into a 14.2 release, my team would really benefit both from this PR

In my opinion, this change is a large enough to be in the major release `0.15` instead of `0.14.2`. The preferred approach for us would be a pre-release `0.15.dev0`. Would this work for you? Note that `datajoint` will continue to point to `datajoint==0.14.1` as the latest, but your team could explicitly set `datajoint==0.15.dev0`.

**Comment by CBroz1** [comment]
Sure, that works for me, thanks! Should I make any edits to the changelog to reflect the pre-release?

**Review by ethho — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,7 @@
 - Fixed - `Autopopulate.populate` excludes `reserved` jobs in addition to `ignore` and `error` jobs 
 - Fixed - Issue [#1159]((https://github.com/datajoint/datajoint-python/pull/1159)  (cascading delete)  - PR [#1160](https://github.com/datajoint/datajoint-python/pull/1160)
```
Comment (ethho) [inline CHANGELOG.md:14]:
```suggestion
- Fixed - Issue [#1159](https://github.com/datajoint/datajoint-python/pull/1159) (cascading delete) - PR [#1160](https://github.com/datajoint/datajoint-python/pull/1160)
```

@CBroz1 Just this change. I need to pivot to other tasks today but will work on publishing `0.15.dev0` tomorrow.

Comment (CBroz1) [reply CHANGELOG.md]: > publishing `0.15.dev0` tomorrow.

Thanks so much!

**Comment by ethho** [comment]
For anyone urgently blocked on `0.15.dev0`, note that you can always `pip install git+`: https://stackoverflow.com/a/20101940

File: tests/test_cascading_delete.py [hunk 3]
```diff
@@ -110,11 +111,24 @@ def test_delete_master(schema_simp_pop):
...
+def test_delete_parts(schema_simp_pop):
+    """test issue #151"""
+    Profile().populate_random()
+    Website().delete(force_masters=True)
+
+
+def test_delete_parts_complex(schema_simp_pop):
+    """test issue #151 with complex master/part. PR #1158."""
+    prev_len = len(G())
+    (A() & "id_a=1").delete()
```
Comment (ethho) [inline tests/test_cascading_delete.py:130]:
```suggestion
    (A() & "id_a=1").delete(force_masters=True)
```

**Review by ethho — CHANGES_REQUESTED** [review]

**Review by ethho — APPROVED** [review]

**Comment by ethho** [comment]
@CBroz1 @edeno @lfrank This PR was included in [release 0.4.2](https://pypi.org/project/datajoint/0.14.2/) built by [CI](https://github.com/datajoint/datajoint-python/actions/runs/10514098880/job/29131408143).
